### PR TITLE
Feat: support android 13 permission

### PR DESCRIPTION
**Pull request by CaiJingLong**

### CODE REVIEW

**Comment by CaiJingLong** [comment]
The main code part is completed.
The question now is whether the method containing `requestType` needs to be verified according to the manifest.

**Comment by AlexV525** [comment]
> The question now is whether the method containing `requestType` needs to be verified according to the manifest.

IMO it must be. Requests with specific types can only be made when corresponding permissions are granted. Such as videos can be fetched when `REQUEST_MEDIA_VIDEO` is granted.

**Comment by CaiJingLong** [comment]
https://github.com/fluttercandies/flutter_photo_manager/blob/af6e6e601035fbf1ab119252f2b9ea64b90b0488/example/android/app/src/main/AndroidManifest.xml#L11-L13

Just remove or add permission item for check the PR code.

File: README.md [hunk 1]
```diff
@@ -525,6 +539,18 @@ rootProject.allprojects {
 See [ProGuard for Glide](https://github.com/bumptech/glide#proguard)
 if you want to know more about using ProGuard and Glide together.
 
+#### Android 13 (Api33) extra configs
```
Comment (AlexV525) [inline README.md:542]:
```suggestion
#### Android 13 (API 33) extra configs
```

File: README.md [hunk 2]
```diff
@@ -33,56 +33,62 @@ For versions upgrade across major versions,
...
+        - [Long retrieving duration with file on iOS](#long-retrieving-duration-with-file-on-ios)
+    - [Entities change notify](#entities-change-notify)
+  - [Cache mechanism](#cache-mechanism)
+    - [Cache on Android](#cache-on-android)
+    - [Cache on iOS](#cache-on-ios)
+    - [Clear caches](#clear-caches)
+  - [Native extra configs](#native-extra-configs)
+    - [Android extra configs](#android-extra-configs)
+      - [Glide issues](#glide-issues)
+      - [Android 13 (Api33) extra configs](#android-13-api33-extra-configs)
```
Comment (AlexV525) [inline README.md:77]:
```suggestion
      - [Android 13 (API 33) extra configs](#android-13-api-33-extra-configs)
```

File: README.md [hunk 3]
```diff
@@ -525,6 +539,18 @@ rootProject.allprojects {
 See [ProGuard for Glide](https://github.com/bumptech/glide#proguard)
 if you want to know more about using ProGuard and Glide together.
 
+#### Android 13 (Api33) extra configs
+
+When you are using Android 13 (API level 33), you need to add the following extra configs to your manifest:
+
+```xml
+<manifest
```
Comment (AlexV525) [inline README.md:547]:
```suggestion
<manifest>
```

File: README.md [hunk 4]
```diff
@@ -525,6 +539,18 @@ rootProject.allprojects {
 See [ProGuard for Glide](https://github.com/bumptech/glide#proguard)
 if you want to know more about using ProGuard and Glide together.
 
+#### Android 13 (Api33) extra configs
+
+When you are using Android 13 (API level 33), you need to add the following extra configs to your manifest:
```
Comment (AlexV525) [inline README.md:544]:
```suggestion
When targeting Android 13 (API level 33), the following extra configs needs to be added to the manifest:
```

File: android/src/main/kotlin/com/fluttercandies/photo_manager/constant/Methods.kt [hunk 5]
```diff
@@ -31,5 +29,19 @@ class Methods {
...
+
+        val android13PermissionMethods =
+            arrayOf(
+                fetchPathProperties,
+                getAssetPathList,
+                getAssetListPaged,
+                getAssetListRange,
+            )
     }
 }
```
Comment (AlexV525) [inline android/src/main/kotlin/com/fluttercandies/photo_manager/constant/Methods.kt:47]: Need a proper EOF.

File: android/src/main/kotlin/com/fluttercandies/photo_manager/constant/Methods.kt [hunk 6]
```diff
@@ -31,5 +29,19 @@ class Methods {
         const val copyAsset = "copyAsset"
         const val moveAssetToPath = "moveAssetToPath"
         const val removeNoExistsAssets = "removeNoExistsAssets"
+
+        /// /// The 4 methods have [RequestType] params, for android-13, it will check permission.
```
Comment (AlexV525) [inline android/src/main/kotlin/com/fluttercandies/photo_manager/constant/Methods.kt:33]:
```suggestion
        /// Below methods have [RequestType] params, thus permissions are required for Android 13.
```

File: android/src/main/kotlin/com/fluttercandies/photo_manager/permission/PermissionsUtils.kt [hunk 7]
```diff
@@ -202,8 +211,62 @@ class PermissionsUtils {
...
+
+    fun havePermissionInManifest(context: Context, permission: String): Boolean {
+        val applicationInfo = context.applicationInfo
+        val packageInfo = context.packageManager.getPackageInfo(
+            applicationInfo.packageName,
+            PackageManager.GET_PERMISSIONS
+        )
+        return packageInfo.requestedPermissions.contains(permission)
+    }
 }
```
Comment (AlexV525) [inline android/src/main/kotlin/com/fluttercandies/photo_manager/permission/PermissionsUtils.kt:272]: Need a proper EOF.

File: lib/src/internal/constants.dart [hunk 8]
```diff
@@ -14,11 +15,14 @@ class PMConstants {
   static const String mRequestPermissionExtend = 'requestPermissionExtend';
   static const String mPresentLimited = 'presentLimited';
   static const String mFetchEntityProperties = 'fetchEntityProperties';
+  static const String mGetAssetCountFromPath = 'getAssetCountFromPath';
+
+  /// The 4 methods have [RequestType] params, for android-13
```
Comment (AlexV525) [inline lib/src/internal/constants.dart:20]:
```suggestion
  /// Below methods have [RequestType] params, thus permissions are required for Android 13.
```

File: example/pubspec.yaml [hunk 9]
```diff
@@ -12,7 +12,7 @@ dependencies:
     sdk: flutter
 
   http: ^0.13.0
-  oktoast: ^3.2.0
+  oktoast: ^3.1.0
```
Comment (AlexV525) [inline example/pubspec.yaml:15]: Is this intended?

File: lib/src/internal/constants.dart [hunk 10]
```diff
@@ -3,6 +3,7 @@
 // in the LICENSE file.
 
 import '../types/thumbnail.dart';
+import '../types/types.dart';
```
Comment (AlexV525) [inline lib/src/internal/constants.dart:6]: This is redundant. Docs will inject type references even if you didn't import them.

File: example/pubspec.yaml (hunk 9 above)
Comment (CaiJingLong) [reply example/pubspec.yaml:15]: My flutter 3.x.x sdk have some error, oktoast 3.2 cannot run on my device ,I think oktoast version in example maybe to 'any'.

File: lib/src/internal/constants.dart (hunk 10 above)
Comment (CaiJingLong) [reply lib/src/internal/constants.dart:6]: In the IDEA/VSCode, if I not import the dart, I can't cmd+click to jump the source to the `RequestType` to the file of defined.

**Review by AlexV525 — APPROVED** [review]

**Pull request merged**